Documentation of ModelQL v2 (#187)
modelix/modelix.core: docs/global/modules/core/pages/explanation/modelql.adoc

= ModelQL

When working with large models you will quickly run into performance issues
when you try to replicate the whole model into the client.

> While the data structure for model replication in Modelix supports partial loading of models

Reviewer: ```suggestion
While the data structure for model replication in Modelix supports partial loading of models,
```

you still need a way to describe which data you need on the client.
Loading data on demand while traversing the model also results in a poor performance,
because of the potentially large number of fine-grained request.

A first attempt to solve this problem was to disallow lazy loading
and require the client to load all required data at the beginning,
before working with the model.

> A special query language was used to filter the data and an attempt to access a node that is not included by that query
> results in an exception, forcing the developer to adjust the query.

Reviewer: ```suggestion
A special query language was used to filter the data and an attempt to access a node that is not included by that query
resulted in an exception, forcing the developer to adjust the query.
```

> While this results in a more predictable performance it's hard to maintain and still not optimal for the performance.

Reviewer: ```suggestion
While this results in a more predictable performance, it is also hard to maintain and still not optimal for the performance.
```

> You have to download all the data at the beginning that you might eventually need.

Reviewer: ```suggestion
You have to download all the data at the beginning that you might eventually need, potentially exceeding the available memory of the system.
```

The ModelQL query language provides a more dynamic way of loading parts of the model on demand,

> but still allows to reduce the number of request to a minimum.

Reviewer: ```suggestion
but still allows reducing the number of request to a minimum.
```

Either "allows <subject> to reduce" or "allows reducing" is valid grammar.

The downside is, that it's not just a different implementation hidden behind the model-api,
but requires to use a different API.

== Reactive Streams

> 

Reviewer: Here's some pretty harsh gap. The section before give a general idea of the problem to solve with ModelQL and the following section is very technical. Some information in between how things are solved would be helpful here.

Reviewer: @slisson what about this one? Skip for later?

> The downside is, that it's not just a different implementation hidden behind the model-api,

Reviewer: ```suggestion
The downside is that it's not just a different implementation hidden behind the model-api,
```
